### PR TITLE
Bug/remove alignment

### PR DESCRIPTION
**Pull request by amitgilad3**

### CODE REVIEW

File: src/scope/models/component.js [hunk 1]
```diff
@@ -205,7 +206,14 @@ export default class Component extends BitObject {
    */
   remove(repo: Repository, deepRemove: boolean = false): Promise {
     const objectRefs = deepRemove ? this.collectRefs(repo) : this.versionArray;
-    return repo.removeMany(objectRefs.concat([this.hash()]));
+    const uniq = uniqBy(objectRefs, 'hash');
```
Comment (GiladShoham) [inline src/scope/models/component.js]: uniq -> uniqRefs

File: src/scope/repositories/sources.js [hunk 2]
```diff
@@ -266,9 +272,19 @@ export default class SourceRepository {
...
   }
 
+  async removeVersion(component, bitId): Promise<void> {
+    const objectRepo = this.objects();
+    const modifiedCompoent = await component.removeVersion(objectRepo, bitId.version);
+    objectRepo.add(modifiedCompoent);
+    await objectRepo.persist();
+    return modifiedCompoent;
+  }
   clean(bitId: BitId, deepRemove: boolean = false): Promise<void> {
```
Comment (GiladShoham) [inline src/scope/repositories/sources.js]: add docs for this function (especially for the deepRemove)

File: src/scope/repositories/sources.js [hunk 3]
```diff
@@ -266,9 +272,19 @@ export default class SourceRepository {
     return component;
   }
 
+  async removeVersion(component, bitId): Promise<void> {
```
Comment (GiladShoham) [inline src/scope/repositories/sources.js]: add docs for the function

File: src/scope/scope.js [hunk 4]
```diff
@@ -700,19 +703,60 @@ export default class Scope {
     return Promise.all(versionDependenciesArr.map(versionDependencies => versionDependencies.toConsumer(this.objects)));
   }
 
+  async removeComponent(id, componentList, removeRefs: boolean = false) {
+    const symlink = componentList.filter(
+      link => link instanceof Symlink && link.id() === id.toStringWithoutScopeAndVersion()
```
Comment (GiladShoham) [inline src/scope/scope.js]: link -> component

File: src/scope/scope.js [hunk 5]
```diff
@@ -700,19 +703,60 @@ export default class Scope {
...
+        await this.removeComponent(id, componentList, true);
+        return id;
+      }
+    });
+
+    return (await Promise.all(removedComponents)).filter(x => !R.isNil(x));
+  }
+
+  /**
+   * Remove  single component
```
Comment (GiladShoham) [inline src/scope/scope.js]: fix docs
add removeSameOrigin

File: src/scope/scope.js [hunk 6]
```diff
@@ -700,19 +703,60 @@ export default class Scope {
...
    * @removeComponent - boolean - true if you want to remove component
    */
-  async removeSingle(bitId: BitId): Promise<string> {
+  async removeSingle(bitId: BitId, removeSameOrigin: boolean = false): Promise<string> {
     logger.debug(`removing ${bitId.toString()}`);
+    const component = (await this.sources.get(bitId)).toComponentVersion();
+    const consumerComponentToRemove = await component.toConsumer(this.objects);
```
Comment (GiladShoham) [inline src/scope/scope.js]: you have a function that does this 2 lines i think

File: src/scope/component-remove.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    removedDependencies: BitIds = [],
+    dependentBits: Object = {}
+  ) {
+    this.removedComponentIds = bitIds;
+    this.missingComponents = missingComponents;
+    this.dependentBits = dependentBits;
+    this.removedDependencies = removedDependencies;
+  }
+
+  paintMissingComponents = () =>
```
Comment (GiladShoham) [inline src/scope/component-remove.js]: I really don't like all the paintings here.. it's class under the scope. it shouldn't be aware of painting at all.
Create a function in the CLI/templates which knows to get a RemovedObjects instance and print it.

File: e2e/commands/remove.e2e.js [hunk 8]
```diff
@@ -121,33 +147,165 @@ describe('bit remove command', function () {
...
+      const output = helper.removeComponent('utils/is-string@0.0.2');
+      expect(output).to.contain.string('successfully removed components: utils/is-string@0.0.2');
+    });
+    it('should display version 0.0.1 for component', () => {
+      const output = JSON.parse(helper.listLocalScope('-j'));
+      expect(output).to.deep.includes({ id: 'utils/is-string', localVersion: '0.0.1' });
+    });
+    it('should still be in bitmap', () => {
+      const bitMap = helper.readBitMap();
+      expect(bitMap).to.have.property('utils/is-string');
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply e2e/commands/remove.e2e.js]: there is no version in bitmap since no export was run

File: src/consumer/consumer.js [hunk 9]
```diff
@@ -1004,32 +1013,132 @@ export default class Consumer {
...
+    return Promise.all(
+      bitIds.map(async (id) => {
+        const component = id.isLocal()
+          ? bitMap.getComponent(bitMap.getExistingComponentId(id.toStringWithoutVersion()))
+          : bitMap.getComponent(id);
+        if (
+          (component.origin && component.origin == COMPONENT_ORIGINS.IMPORTED) ||
+          component.origin == COMPONENT_ORIGINS.NESTED
+        ) {
+          return await fs.remove(path.join(this.getPath(), component.rootDir));
```
Comment (GiladShoham) [inline src/consumer/consumer.js]: you don't need the await here

File: src/consumer/consumer.js [hunk 10]
```diff
@@ -1004,32 +1013,132 @@ export default class Consumer {
...
+        } else if (component.origin == COMPONENT_ORIGINS.AUTHORED && deleteFiles) {
+          return Promise.all(component.files.map(async file => await fs.remove(file.relativePath)));
```
Comment (GiladShoham) [inline src/consumer/consumer.js]: you don't need the async await here

**Pull request merged**